File: app/middleware/error_handler.py
"""
Error Handling Middleware

Provides consistent error handling across all API endpoints.
Catches common HTTP exceptions and returns standardized JSON error responses.
"""

import uuid
from datetime import datetime
from fastapi import Request, status
from fastapi.responses import JSONResponse
from fastapi.exceptions import RequestValidationError
from starlette.exceptions import HTTPException as StarletteHTTPException


async def http_exception_handler(request: Request, exc: StarletteHTTPException) -> JSONResponse:
    """
    Handler for HTTP exceptions (404, 403, etc.).

    Converts standard HTTP exceptions into consistent JSON error responses
    with error codes, messages, and tracking information.

    Args:
        request: The incoming request
        exc: The HTTP exception

    Returns:
        JSONResponse: Standardized error response
    """
    error_code_map = {
        status.HTTP_400_BAD_REQUEST: "BAD_REQUEST",
        status.HTTP_401_UNAUTHORIZED: "UNAUTHORIZED",
        status.HTTP_403_FORBIDDEN: "FORBIDDEN",
        status.HTTP_404_NOT_FOUND: "NOT_FOUND",
        status.HTTP_405_METHOD_NOT_ALLOWED: "METHOD_NOT_ALLOWED",
        status.HTTP_409_CONFLICT: "CONFLICT",
        status.HTTP_422_UNPROCESSABLE_ENTITY: "UNPROCESSABLE_ENTITY",
        status.HTTP_429_TOO_MANY_REQUESTS: "TOO_MANY_REQUESTS",
        status.HTTP_500_INTERNAL_SERVER_ERROR: "INTERNAL_SERVER_ERROR",
        status.HTTP_503_SERVICE_UNAVAILABLE: "SERVICE_UNAVAILABLE",
    }

    error_code = error_code_map.get(exc.status_code, "UNKNOWN_ERROR")

    return JSONResponse(
        status_code=exc.status_code,
        content={
            "error": exc.detail if isinstance(exc.detail, str) else str(exc.detail),
            "error_code": error_code,
            "message": exc.detail if isinstance(exc.detail, str) else str(exc.detail),
            "details": None,
            "path": str(request.url.path),
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "request_id": str(uuid.uuid4())
        }
    )


async def validation_exception_handler(request: Request, exc: RequestValidationError) -> JSONResponse:
    """
    Handler for request validation errors (422).

    Formats Pydantic validation errors into a consistent structure
    with detailed information about which fields failed validation.

    Args:
        request: The incoming request
        exc: The validation error

    Returns:
        JSONResponse: Standardized validation error response
    """
    errors = []
    for error in exc.errors():
        errors.append({
            "field": ".".join(str(loc) for loc in error["loc"]),
            "message": error["msg"],
            "type": error["type"]
        })

    return JSONResponse(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        content={
            "error": "Validation error",
            "error_code": "VALIDATION_ERROR",
            "message": "Request validation failed. Please check the request parameters and body.",
            "details": {
                "validation_errors": errors
            },
            "path": str(request.url.path),
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "request_id": str(uuid.uuid4())
        }
    )


async def generic_exception_handler(request: Request, exc: Exception) -> JSONResponse:
    """
    Handler for unhandled exceptions (500).

    Catches any exceptions that weren't handled by other handlers
    and returns a generic error response. In production, detailed
    error messages should be logged but not exposed to clients.

    Args:
        request: The incoming request
        exc: The unhandled exception

    Returns:
        JSONResponse: Generic internal server error response
    """
    # In production, log the full exception details
    # but only return a generic message to the client
    error_message = "An internal server error occurred. Please try again later."

    # In development/debug mode, include more details
    from app.config import settings
    if settings.debug:
        error_message = f"Internal server error: {str(exc)}"

    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content={
            "error": "Internal server error",
            "error_code": "INTERNAL_SERVER_ERROR",
            "message": error_message,
            "details": None,
            "path": str(request.url.path),
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "request_id": str(uuid.uuid4())
        }
    )

File: src/middleware/errorHandler.ts
import { Request, Response, NextFunction } from 'express';
import { PDFExtractionError } from '../utils/errors';
import { logger } from '../utils/logger';

/**
 * Error response structure
 */
interface ErrorResponse {
  error: {
    name: string;
    message: string;
    code?: string;
    details?: any;
    stack?: string;
  };
}

/**
 * Map error codes to HTTP status codes
 */
const ERROR_STATUS_MAP: Record<string, number> = {
  PDF_NOT_FOUND: 404,
  PDF_PERMISSION_DENIED: 403,
  PDF_CORRUPTED: 400,
  PDF_ENCRYPTED: 400,
  PDF_SCANNED: 400,
  PDF_INVALID_FORMAT: 400,
  PDF_INVALID_INPUT: 400,
  PDF_PARSING_FAILED: 422,
  PDF_FILESYSTEM_ERROR: 500,
};

/**
 * Get HTTP status code for error
 */
function getStatusCode(error: any): number {
  // Check if error is a PDFExtractionError with a code
  if (error instanceof PDFExtractionError && error.code) {
    return ERROR_STATUS_MAP[error.code] || 500;
  }

  // Handle common HTTP errors
  if (error.status) {
    return error.status;
  }

  if (error.statusCode) {
    return error.statusCode;
  }

  // Default to 500 for unknown errors
  return 500;
}

/**
 * Build error response object
 */
function buildErrorResponse(error: any, includeStack: boolean): ErrorResponse {
  const response: ErrorResponse = {
    error: {
      name: error.name || 'Error',
      message: error.message || 'An unexpected error occurred',
    },
  };

  // Add error code if available
  if (error.code) {
    response.error.code = error.code;
  }

  // Add error details if available
  if (error.details) {
    response.error.details = error.details;
  }

  // Include stack trace in development
  if (includeStack && error.stack) {
    response.error.stack = error.stack;
  }

  return response;
}

/**
 * Express error handling middleware
 *
 * Catches all errors thrown in request handlers and formats them
 * as JSON responses with appropriate HTTP status codes.
 *
 * @example
 * ```typescript
 * app.use(errorHandler);
 * ```
 */
export function errorHandler(
  error: any,
  req: Request,
  res: Response,
  next: NextFunction
): void {
  // If headers already sent, delegate to default Express error handler
  if (res.headersSent) {
    return next(error);
  }

  const statusCode = getStatusCode(error);
  const isDevelopment = process.env.NODE_ENV === 'development';

  // Log the error
  logger.error(
    `${req.method} ${req.path} - ${error.message}`,
    error
  );

  // Build and send error response
  const errorResponse = buildErrorResponse(error, isDevelopment);
  res.status(statusCode).json(errorResponse);
}

/**
 * Catch-all middleware for 404 Not Found errors
 *
 * Should be registered after all routes.
 *
 * @example
 * ```typescript
 * app.use(notFoundHandler);
 * ```
 */
export function notFoundHandler(
  req: Request,
  _res: Response,
  next: NextFunction
): void {
  const error = {
    name: 'NotFoundError',
    message: `Route ${req.method} ${req.path} not found`,
    code: 'ROUTE_NOT_FOUND',
    status: 404,
  };

  next(error);
}
